File: code/xproblem/Java/xproblem/visitorPattern/Evaluator.java
package visitorPattern;

/**
 * Operation for expression evaluation
 */
public class Evaluator extends Visitor<Integer> {
	public Integer visit(Const x) {
		return x.getInfo();
	}
	public Integer visit(Add x) {
		return x.getLeft().accept(this) + x.getRight().accept(this);
	}
	public Integer visit(Neg x) {
		return - x.getExpr().accept(this);
	}
}

File: code/xproblem/CSharp/xproblem/visitorPattern/Evaluator.cs
namespace visitorPattern
{
    public class Evaluator : Visitor<int>
    {
        public int Visit(Const that)
        {
            return that.info;
        }
        public int Visit(Add that)
        {
            return that.left.Accept(this) + that.right.Accept(this);
        }
    }
}
